src/citrine/resources/ara_definition.py

```diff
@@ -124,6 +126,74 @@ def add_columns(self, *,
             columns=copy(self.columns) + columns
         )
 
+    def add_all_ingredients(self, *,
+                            process_template: LinkByUID,
+                            project,
+                            quantity_dimension: IngredientQuantityDimension,
+                            scope: str = 'id'
+                            ):
+        """[ALPHA] Add variables and columns for all of the possible ingredients in a process.
+
+        For each allowed ingredient name in the process template there is a column for the if of
+        the ingredient and a column for the quantity of the ingredient. If the quantities are
+        given in absolute amounts then there is also a column for units.
+
+        Parameters
+        ------------
+        process_template: LinkByUID
+            scope and id of a registered process template
+        project: Project
+            a project that has access to the process template
+        quantity_dimension: IngredientQuantityDimension
+            the dimension in which to report ingredient quantities
+        scope: Optional[str]
+            the scope for which to get ingredient ids (default is Citrine scope, 'id')
+
+        """
+        quantity_dimension_display = {
+            IngredientQuantityDimension.ABSOLUTE: "absolute quantity",
+            IngredientQuantityDimension.MASS: "mass fraction",
+            IngredientQuantityDimension.VOLUME: "volume fraction",
+            IngredientQuantityDimension.NUMBER: "number fraction"
+        }
+        process: ProcessTemplate = project.process_templates.get(
+            uid=process_template.id, scope=process_template.scope)
+        new_variables = []
+        new_columns = []
+        for name in process.allowed_names:
+            identifier_variable = IngredientIdentifierByProcessTemplateAndName(
+                name="{} of ingredient {} in process {}".format(scope, name, process.name),
```

Comment:
I'm worried about accidental collisions with other names (either if the user adds them manually or if there is a different process template that happens to have the same name and includes an ingredient of the same name). Should we include a unique ID of some sort here, such as the id of the process template?

Reply:
`name` is never user visible.  It would make sense to use some kind of hash function or randomized string.

Reply:
I want it to be deterministic because I want to prevent the user from adding the same column twice. I think a hash based on the ingredient and process name will work.

Reply:
A hash based on the process UUID and the ingredient name would be nearly guaranteed unique.

Reply:
Done. I also kept in the strings themselves, just in case someone in the future is debugging something on platform-backend and having some human-readable pieces of the variable names makes their lives easier.